### DISCUSSION TITLE
Any difference between COPY and read_csv?

### DISCUSSION THREAD

**Discussion by goodwanghan** [duckdb/duckdb#2892] [Q&A]
Hello,

For CSV loading, we see two options from [here](https://duckdb.org/docs/data/csv) and [here](https://duckdb.org/docs/sql/statements/copy).

So I have a few questions:

1. Does `read_csv_auto`/`read_csv` have lazy evaluation, and how about `COPY`?
2. I couldn't find the API docs for `read_csv_auto`/`read_csv` I think their parameter names are slightly different from `COPY`?
3. I also encountered [this issue](https://github.com/duckdb/duckdb/issues/2641) and see it's fixed in pre-release and I wonder when will the next release happen?

Thanks!
Han

**Comment by Mytherin** [comment]
`read_csv` and `COPY` are the same, except that `COPY` can only be used for a single purpose (copy data from a CSV directly into a table). `read_csv` is more flexible in that it can be used inside queries, meaning you can query CSVs directly, or do additional preprocessing on the contents of the CSV file prior to inserting data into a table. They use the exact same CSV reader functionality internally. `read_csv` documentation can be found [here](https://duckdb.org/docs/data/csv).

CSV reading is done in a streaming manner (perhaps that is what you mean with lazy evaluation?). The CSV reader reads chunks of 1024 tuples at a time, meaning if you read from a CSV file with e.g. a `LIMIT` clause only the first few tuples will be read.

The next release will likely happen sometime this month.

**Reply by goodwanghan** [reply]
Thanks @Mytherin , But [this](https://duckdb.org/docs/data/csv) doc does not list all parameters of `read_csv` and `read_csv_auto`.

It's great to know `COPY` is the same, I will use that, where I can find the doc at [postgressql](https://www.postgresql.org/docs/13/sql-copy.html)

**Reply by Mytherin** [reply]
Is there anything wrong with the list of parameters in [our COPY docs](https://duckdb.org/docs/sql/statements/copy)? While many parameters are shared with Postgres we do have additional ones there.

**Reply by goodwanghan** [reply]
Oh right, by using `COPY` I can read your COPY docs.

Would it be possible to also have the similar doc for `read_csv` and `read_csv_auto` to list all parameters?

**Reply by Mytherin** [reply]
Yes, this should be added. I will have a go at it.

**Reply by goodwanghan** [reply]
Thanks, also does `read_csv` support file array like `parquet_scan`?

**Reply by Mytherin** [reply]
Documentation is [here now](https://duckdb.org/docs/data/csv#parameters). File arrays are not yet supported, but should be rather easy to add. Feel free to open a feature request.

**Reply by goodwanghan** [reply]
That is fast! Thank you!

**Comment by Alex-Monahan** [comment]
Also, which DuckDB client are you using? There are ways to use the master branch version in the meantime!

**Reply by goodwanghan** [reply]
`\t` issues is in 0.3.1, but in the pre-release version, it's already fixed
